operator/pkg/ciliumidentity/controller.go

```diff
@@ -108,15 +134,137 @@ func (c *Controller) Start(_ cell.HookContext) error {
+	// Processing Pod work queue items are adding items to CID work queue.
+	// Processed CID work queue items result in mutations to CID API objects.
+	//
+	// Diagram:
+	//-------------------------Pod,CID event
+	//---------------------------||
+	//----------------------------V
+	// Namespace event -> Resource work queue -> Mutate CID API objects
+	c.startEventProcessing()
+	c.startWorkQueues()
+
 	return nil
 }
 
-func (c *Controller) enqueueCIDReconciliation(cidKey resource.Key, delay time.Duration) {
+func (c *Controller) Stop(_ cell.HookContext) error {
+	c.resourceQueue.ShutDown()
+	c.contextCancel()
+
+	return nil
+}
+
+func (c *Controller) initializeQueues() {
+
+	c.logger.Info("CID controller resource work queue configuration",
+		logfields.WorkQueueSyncBackOff, defaultSyncBackOff,
+		logfields.WorkQueueMaxSyncBackOff, maxSyncBackOff)
+
+	c.resourceQueue = workqueue.NewRateLimitingQueueWithConfig(
+		workqueue.NewItemExponentialFailureRateLimiter(defaultSyncBackOff, maxSyncBackOff),
+		workqueue.RateLimitingQueueConfig{Name: "ciliumidentity_resource"})
+}
+
+func (c *Controller) startEventProcessing() {
+
+	c.jobGroup.Add(
+		job.OneShot("proc-cid-events", func(ctx context.Context, health cell.Health) error {
+			return c.processCiliumIdentityEvents(ctx)
+		}),
+
+		job.OneShot("proc-pod-events", func(ctx context.Context, health cell.Health) error {
+			return c.processPodEvents(ctx)
+		}),
+
+		job.OneShot("proc-ces-events", func(ctx context.Context, health cell.Health) error {
+			return c.processCiliumEndpointSliceEvents(ctx)
+		}),
+
+		job.OneShot("proc-ns-events", func(ctx context.Context, health cell.Health) error {
+			return c.processNamespaceEvents(ctx)
+		}),
+	)
 }
 
-func (c *Controller) enqueuePodReconciliation(podKey resource.Key, delay time.Duration) {
+func (c *Controller) startWorkQueues() {
+
+	c.jobGroup.Add(
+		job.OneShot("op-managing-resource-wq", func(ctx context.Context, health cell.Health) error {
+			return c.runResourceWorker(ctx)
+		}),
+	)
 }
 
-func (c *Controller) Stop(hookContext cell.HookContext) error {
+func (c *Controller) runResourceWorker(context context.Context) error {
+	c.logger.Info("Starting resource worker")
+	defer c.logger.Info("Stopping resource worker")
+
+	for c.processNextItem() {
+		select {
+		case <-context.Done():
+			return nil
+		case <-c.context.Done():
+			return nil
+		default:
+		}
+	}
+
 	return nil
 }
+
+func (c *Controller) enqueueReconciliation(item QueuedItem, delay time.Duration) {
+	c.resourceQueue.AddAfter(item, delay)
+}
+
+func (c *Controller) processNextItem() bool {
+	item, quit := c.resourceQueue.Get()
+	if quit {
+		return false
+	}
+	defer c.resourceQueue.Done(item)
+
+	qItem := item.(QueuedItem)
+	key := qItem.Key()
+
+	var err error
+	switch qItem.(type) {
+	case PodItem:
+		err = c.reconciler.reconcilePod(key)
+	case CIDItem:
+		err = c.reconciler.reconcileCID(key)
+	}
+
+	if err != nil {
+		c.handleErr(err, qItem)
+	}
```

Comment:
If we extend the `QueuedItem` interface with:

```go
  Reconcile(reconciler *reconciler) error
```

and add those methods to `PodItem` and `CIDItem`, respectively:

```go
func (p PodItem) Reconcile(reconciler *reconciler) error {
       return reconciler.reconcilePod(p.key)
}

func (c CIDItem) Reconcile(reconciler *reconciler) error {
       return reconciler.reconcileCID(c.key)
}
```

then the snippet above can be simplified with:

```go
        if err := qItem.Reconcile(c.reconciler); err != nil {
                c.handleErr(err, qItem)
        }
```

Not a big difference, but I find this a bit cleaner. WDYT?